_posts/2020-04-11-native-mobile-to-flutter.markdown

```diff
@@ -0,0 +1,122 @@
+- Testing utilities are definitely better on React.
+- More third party library support.
+
+#### Things we liked about Flutter:
+
+- Without javascript developers on our mobile team, the learning curve for Dart/Flutter was substantially lower than Javascript/React Native. 
+- Flutter offered clean, straight-forward conventions that were easy to learn, expressive and explicit.
+- Android Studio offered a turnkey development environment with performance monitoring tools, hot reload and step through debugging and inspection.  Since most of our developers are Android developers, the IDE was an added bonus due to its familiarity.
+- Flutter does not rely on any native UI components, so it's largely immune to breaking changes introduced by Apple, Google, or Android device manufacturers.
+
+### Choosing a State Management pattern that matched our needs.
+
+There were several state management architectures to choose from, but we opted for simplicity, the [Scoped Model pattern](https://github.com/brianegan/scoped_model).  The scoped model pattern is a simple change notifier pattern bound to Flutter's UI widget hierarchy.  It has gradually been superceded in the community by the [Provider library](https://github.com/rrousselGit/provide) which is based on Scoped Model.  We will most likely be migrating to the provider library within the next year.
+
+We use this pattern with our SQLite database to allow sql table subscriptions that trigger redraws on subscribed user interfaces when data is modified on disk.  We also use a very simple stream subscription pattern to propagate state changes that are not tied to database state.  Flutter provides the excellent [StreamBuilder class](https://api.flutter.dev/flutter/widgets/StreamBuilder-class.html) to facilitate state propagation.
+
+We evaluated a few other patterns mentioned in flutter resources, particularly Redux and BLoC, but chose the simplicity of Scoped Model over the granularity and associated complexity that the others presented.  
+
+### Native integrations. 
+
+Flutter offers a very simple message passing api for communicating with native code.  We needed to use native integrations for a relatively small group of features: 
+
+- Push notification.
+- Deep link handling.
+- Connectivity detection.
+- Security features.
+- Device information. 
+
+_This platform specific code amounted to less than 2% of our total lines of code._ 
+
+We were able to write this functionality easily and communicate between Flutter and native with very little code overhead. Flutter provides a [very simple api](https://flutter.dev/docs/development/platform-integration/platform-channels#architecture) for communicating bidirectionally between native and dart code.
+
+### Unit tests.
+
+Unit testing in Flutter is straightforward, but lacks the robust testing utilities and third party libraries found in React.  We rolled most of our own testing scaffolding aside from the simple test runner offered by flutter.
+
+### UI testing. 
```

Comment:
Would be helpful to clarify if we're talking about true end to end UI testing, where our mounted widgets actually hit networked services or if we're limiting our UI tests to clicking around the app with a stubbed/mocked API.

Also would be nice to expand a bit on the reason for test flakiness - e.g. "the widget structure was always changing so our tests would break because they couldn't find the right test id" or whatever.